File: java/303_Range_Sum_Query-Immutable.java
/*
https://leetcode.com/problems/range-sum-query-immutable/
Given an integer array nums, find the sum of the elements between indices i and j (i ≤ j), inclusive.
You may assume that the array does not change.  There are many calls to sumRange function.

Store the cumulative sum up to each index.  sumRange is the difference of the cumulative sums to j and to i-1.
Time - O(n) to initialise, O(1) for prefixSums()
Space - O(n)
*/

public class NumArray {

    private int[] prefixSums;

    public NumArray(int[] nums) {
        prefixSums = new int[nums.length + 1];  // prefixSums[i+1] is cumulative sum including nums[i]
        for (int i = 0; i < nums.length; ++i)
            prefixSums[i+1] = nums[i] + prefixSums[i];
    }

    public int sumRange(int i, int j) {
        return prefixSums[j+1] - prefixSums[i];
    }
}

File: python_1_to_1000/303_Range_Sum_Query_-_Immutable.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/range-sum-query-immutable/
# Given an integer array nums, find the sum of the elements between indices i and j (i ≤ j), inclusive.

# Store cumulative sums.
# Time - O(n) for __init__(), O(1) for sumRange()
# Space - O(n)

class NumArray(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.cumul = [0]
        for num in nums:
            self.cumul.append(self.cumul[-1] + num)

    def sumRange(self, i, j):
        """
        :type i: int
        :type j: int
        :rtype: int
        """
        return self.cumul[j + 1] - self.cumul[i]